packages/state-transition/src/block/processWithdrawals.ts

```diff
@@ -95,7 +97,19 @@ export function getExpectedWithdrawals(
   if (fork >= ForkSeq.electra) {
     const stateElectra = state as CachedBeaconStateElectra;
 
-    for (const withdrawal of stateElectra.pendingPartialWithdrawals.getAllReadonly()) {
+    // MAX_PENDING_PARTIALS_PER_WITHDRAWALS_SWEEP = 8, PENDING_PARTIAL_WITHDRAWALS_LIMIT: 134217728 so we should only call getAllReadonly() if it makes sense
+    // pendingPartialWithdrawals comes from EIP-7002 smart contract where it takes fee so it's more likely than not validator is in correct condition to withdraw
+    // also we may break early if withdrawableEpoch > epoch
+    const allPendingPartialWithdrawals =
+      stateElectra.pendingPartialWithdrawals.length <= MAX_PENDING_PARTIALS_PER_WITHDRAWALS_SWEEP
+        ? stateElectra.pendingPartialWithdrawals.getAllReadonly()
+        : null;
```

Comment:
that'd actually make it worse than the initial version because `getAllReadonly()` is way more performant than calling `getReadonly(i)` n times especially if n is big